### PR TITLE
NC | Bucket Owner Removal

### PR DESCRIPTION
**Pull request by romayalon**
### Disclaimers - 
1. This PR is based on the bucket owner removal part of https://github.com/noobaa/noobaa-core/pull/8167 written by @alphaprinz.
2. ~~This PR is dependent on merging - https://github.com/noobaa/noobaa-core/pull/8279 and currently, cherry-picking its only commit, which will be removed after its merge.~~

### Explain the changes
1. Schema - Removal of bucket_owner property requirement from nsfs_bucket_schema.js. Didn't remove it completely from the schema for backward compatibility.
2. CLI - Create/Update flows - removal of bucket_owner property usage from the schema, but left it on the CLI API which means that on create/update bucket `--owner` flag is still the name of the account, and it'll be mapped to an id internally - owner_account.
3. S3 API (BucketspaceFS) - 
- read_bucket_sdk_info() - since we don't keep the bucket owner name we now read the account config file by its id (owner account) and set the bucket_owner run time object.
- create_bucket() - removed bucket_owner from new_bucket_defaults.
4. IAM API (AccountspaceFS) - 
- _check_if_root_account_does_not_have_buckets_before_deletion() - Updated usage by id instead of by name.  
6. Tests/Docs - Removed and updated bucket_owner related changes.

### Issues: Fixed #xxx / Gap #xxx
1. Fixed #https://github.com/noobaa/noobaa-core/issues/8080
2. Gap - we need to remove [the following line](https://github.com/noobaa/noobaa-core/pull/8167/files#diff-11cc61806e3f6607fe678e081751019f21db0f09d48a56a26713203b3ada1db5L241) when having IAM users in users/ directory (this will mean that a root account can have the same name as an IAM user) @shirady 

```diff
-   if (account_identifier === bucket_owner.unwrap()) return true;
+   if (owner_account && account._id === owner_account.id) return true;
+   //name check - only for root accounts
+   if (!account.owner && account_identifier_name === bucket_owner.unwrap()) return true;
```
### Testing Instructions:
1. 

- [ ] Doc added/updated
- [ ] Tests added

### CODE REVIEW

File: src/test/unit_tests/jest_tests/test_config_fs.test.js [hunk 1]
```diff
@@ -22,9 +22,10 @@ describe('adjust_bucket_with_schema_updates', () => {
     });
 
     it('should return bucket without deprecated properties (system_owner)', () => {
```
Comment (shirady) [inline src/test/unit_tests/jest_tests/test_config_fs.test.js:24]: Please update the `it`
```suggestion
    it('should return bucket without deprecated properties', () => {
```

File: src/sdk/bucketspace_fs.js [hunk 2]
```diff
@@ -17,6 +17,7 @@ const nsfs_schema_utils = require('../manage_nsfs/nsfs_schema_utils');
 const nc_mkm = require('../manage_nsfs/nc_master_key_manager').get_instance();
 const { ConfigFS, JSON_SUFFIX } = require('./config_fs');
 const mongo_utils = require('../util/mongo_utils');
+const { TYPES } = require('../manage_nsfs/manage_nsfs_constants');
```
Comment (shirady) [inline src/sdk/bucketspace_fs.js:20]: Please use `CONFIG_TYPES` instead.

File: src/sdk/bucketspace_fs.js [hunk 3]
```diff
@@ -121,11 +122,22 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
             };
 
             bucket.name = new SensitiveString(bucket.name);
-            bucket.bucket_owner = new SensitiveString(bucket.bucket_owner);
+            // Backward compatibility will take longer without bucket owner name
```
Comment (shirady) [inline src/sdk/bucketspace_fs.js:125]: Why do you add this comment "will take longer without bucket owner name"?
There is no other option that we can use...

File: src/sdk/bucketspace_fs.js [hunk 4]
```diff
@@ -121,11 +122,22 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
...
+            const account_config = await this.config_fs.get_identity_by_id(
+                bucket.owner_account,
+                TYPES.ACCOUNT,
+                { silent_if_missing: true }
+            );
+
+            if (!account_config) {
+                dbg.warn(`Bucket Owner does not exist ${bucket.owner_account}`);
+            }
+            bucket.bucket_owner = new SensitiveString(account_config?.name);
```
Comment (shirady) [inline src/sdk/bucketspace_fs.js:135]: if the `account_config` is undefined then you wrap the undefined value, is it what you want?

File: src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js [hunk 5]
```diff
@@ -511,7 +511,8 @@ describe('manage nsfs cli bucket flow', () => {
             await set_path_permissions_and_owner(bucket_defaults.path, account_defaults2, 0o700);
             await exec_manage_cli(TYPES.BUCKET, action, bucket_options);
             const bucket = await config_fs.get_bucket_by_name(bucket_defaults.name);
-            expect(bucket.bucket_owner).toBe(account_defaults2.name);
+            const owner_data = await config_fs.get_identity_by_id(bucket.owner_account);
```
Comment (shirady) [inline src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js:514]: Why did you change it from "bucket_owner" to "owner_data"?

File: src/test/unit_tests/test_bucketspace.js [hunk 6]
```diff
@@ -373,8 +373,8 @@ mocha.describe('bucket operations - namespace_fs', function() {
         const s3_bucket = bucket_name + '-s3';
         const res = await s3_correct_uid_default_nsr.getBucketAcl({ Bucket: s3_bucket });
         const bucket_info = await rpc_client.bucket.read_bucket({ name: s3_bucket });
-        assert.equal(res.Owner.DisplayName, bucket_info.owner_account.email);
         assert.equal(res.Owner.ID, bucket_info.owner_account.id);
+        assert.equal(res.Owner.DisplayName, bucket_info.owner_account.email);
```
Comment (shirady) [inline src/test/unit_tests/test_bucketspace.js:377]: Does this change have a meaning?
If not, then better restore this file.

File: src/manage_nsfs/manage_nsfs_cli_utils.js [hunk 7]
```diff
@@ -38,14 +38,17 @@ function write_stdout_response(response_code, detail, event_arg) {
  * otherwise it would throw an error
  * @param {import('../sdk/config_fs').ConfigFS} config_fs
  * @param {string} bucket_owner
```
Comment (shirady) [inline src/manage_nsfs/manage_nsfs_cli_utils.js:40]: also optional:
```suggestion
 * @param {string} [bucket_owner]
```

File: src/cmd/manage_nsfs.js [hunk 8]
```diff
@@ -96,8 +96,6 @@ async function fetch_bucket_data(action, user_input) {
         // added undefined values to keep the order the properties when printing the data object
         _id: undefined,
         name: user_input.name === undefined ? undefined : String(user_input.name),
-        owner_account: undefined,
```
Comment (shirady) [inline src/cmd/manage_nsfs.js:99]: I think we had this "owner_account" on purpose to have the property saved as undfined.

File: src/sdk/bucketspace_fs.js (hunk 3 above)
Comment (romayalon) [reply src/sdk/bucketspace_fs.js:125]: there is, but it might cause a mistake, I thought of using the old bucket_owner property, read the account config file and check if it has the same id...

File: src/sdk/bucketspace_fs.js (hunk 4 above)
Comment (romayalon) [reply src/sdk/bucketspace_fs.js:135]: initially I threw an error if the owner account config does not exist, but later on I saw it doesn't make sense to fail on that, so since we expect a sensitive string I need to wrap something.. do you have a better suggestion?

File: src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js (hunk 5 above)
Comment (romayalon) [reply src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js:514]: because it's the owner data, bucket_owner IMO should be the bucket owner name as we use on bucketspaceFS

File: src/cmd/manage_nsfs.js [hunk 9]
```diff
@@ -194,6 +196,8 @@ async function update_bucket(data) {
         // for validating against the schema we need an object, hence we parse it back to object
         nsfs_schema_utils.validate_bucket_schema(JSON.parse(data));
         await config_fs.update_bucket_config_file(cur_name, data);
+        data = JSON.parse(data);
```
Comment (shirady) [inline src/cmd/manage_nsfs.js:199]: You can parse once and save it in a variable.

File: src/cmd/manage_nsfs.js [hunk 10]
```diff
@@ -606,17 +612,22 @@ async function list_config_files(type, wide, show_secrets, filters = {}) {
         entries = await config_fs.list_buckets();
```
Comment (shirady) [inline src/cmd/manage_nsfs.js:612]: I would rename it also (less confusing) to `entries_names`

File: src/cmd/manage_nsfs.js [hunk 11]
```diff
@@ -653,6 +664,15 @@ function get_access_keys(action, user_input) {
     return { access_keys, new_access_key };
 }
 
+/**
+ * set_bucker_owner gets bucket owner from cache by its id and sets bucket_owner name on the bucket data
+ * @param {object} bucket_data 
+ */
+async function set_bucker_owner(bucket_data) {
```
Comment (shirady) [inline src/cmd/manage_nsfs.js:671]: A couple of questions related to the account_cache:
1. Now we should use the CLI with a working server in the background? (before that we didn't have to)
2. The cache it per endpoint, and could be many endpoints, how does it work?
3. What happens if the user updates the bucket owner? (`sudo node src/cmd/manage_nsfs bucket update --name <bucket name> --owner <another_account>`) will he see the previous one?
4. Why did you decide to use it in all the operations and not just in bucket list?

File: src/sdk/accountspace_fs.js [hunk 12]
```diff
@@ -26,6 +27,18 @@ const { account_cache } = require('./object_sdk');
 const dummy_region = 'us-west-2';
 const dummy_service_name = 's3';
 
+const account_id_cache = new LRUCache({
```
Comment (shirady) [inline src/sdk/accountspace_fs.js:30]: We already had imported the `account_cache` from object sdk, what is the difference between `account_id_cache` and `account_cache`?

File: src/sdk/bucketspace_fs.js (hunk 3 above)
Comment (shirady) [reply src/sdk/bucketspace_fs.js:124]: I didn't understand this yet.

File: src/sdk/bucketspace_fs.js (hunk 4 above)
Comment (shirady) [reply src/sdk/bucketspace_fs.js:134]: No, I want to make sure this is what you planned.
I suggest to add a comment in the code about this.

File: src/test/unit_tests/test_nc_nsfs_cli.js [hunk 13]
```diff
@@ -20,17 +20,17 @@ const { TYPES, ACTIONS } = require('../../manage_nsfs/manage_nsfs_constants');
 const tmp_fs_path = path.join(TMP_PATH, 'test_bucketspace_fs');
 const DEFAULT_FS_CONFIG = get_process_fs_context();
 const config_fs_account_options = { show_secrets: true, decrypt_secret_key: true };
+const root_path = path.join(tmp_fs_path, 'root_path_manage_nsfs/');
+const config_root = path.join(tmp_fs_path, 'config_root_manage_nsfs');
+const config_fs = new ConfigFS(config_root);
+// TODO: needed for NC_CORETEST FLOW - should be handled better
+const nc_coretes_location = config.NC_MASTER_KEYS_FILE_LOCATION;
```
Comment (shirady) [inline src/test/unit_tests/test_nc_nsfs_cli.js:27]: typo in the variable name, and I would add more details in the name.
```suggestion
const nc_coretest_master_key_location = config.NC_MASTER_KEYS_FILE_LOCATION;
```

File: src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js (hunk 5 above)
Comment (shirady) [reply src/test/unit_tests/jest_tests/test_nc_nsfs_bucket_cli.test.js:514]: We use the `bucket_owner` as a tern of the owner's name only, and `owner data` and the account (objects) which is the owner of the bucket?
I'm fine with it, I think a short comment can help.

**Comment by shirady** [comment]
@romayalon could you please update the the PR description (and attach the issue)?

File: src/sdk/accountspace_fs.js (hunk 12 above)
Comment (romayalon) [reply src/sdk/accountspace_fs.js:30]: The difference is -  
 - Regular accounts cache - access_key -> account mapping
 - new Accounts ID Cache id -> account mapping.

File: src/cmd/manage_nsfs.js [hunk 14]
```diff
@@ -592,24 +604,35 @@ async function list_config_files(type, wide, show_secrets, filters = {}) {
...
             // remove secrets on !show_secrets && should filter
-            return wide ? _.omit(data, show_secrets ? [] : ['access_keys']) : { name: entry };
+            if (!wide) return { name: entry_name };
+            if (type === TYPES.ACCOUNT) return _.omit(data, show_secrets ? [] : ['access_keys']);
+            if (type === TYPES.BUCKET) {
+                data.bucket_owner = bucket_owners_map[data.owner_account];
+                if (!data.bucket_owner) {
+                    await set_bucker_owner(data);
+                    bucket_owners_map[data.owner_account] = data.bucket_owner;
+                }
```
Comment (shirady) [inline src/cmd/manage_nsfs.js:631]: I think that reusing the `bucket_owner` as it was saved in the bucket config will not be valid in case the account was renamed, so I'm not sure that you can really reduce the system call (to read the account config).

File: src/manage_nsfs/manage_nsfs_cli_utils.js [hunk 15]
```diff
@@ -37,15 +37,20 @@ function write_stdout_response(response_code, detail, event_arg) {
  * get_bucket_owner_account will return the account of the bucket_owner
  * otherwise it would throw an error
  * @param {import('../sdk/config_fs').ConfigFS} config_fs
- * @param {string} bucket_owner
+ * @param {string} [bucket_owner]
+ * @param {string} [owner_account_id]
  */
-async function get_bucket_owner_account(config_fs, bucket_owner) {
+async function get_bucket_owner_account(config_fs, bucket_owner, owner_account_id) {
```
Comment (shirady) [inline src/manage_nsfs/manage_nsfs_cli_utils.js:43]: I thought that you are using it only with `bucket_owner` as the name and could remove the argument `owner_account_id` and simplify the function.

File: src/manage_nsfs/manage_nsfs_validations.js [hunk 16]
```diff
@@ -316,7 +316,7 @@ async function validate_bucket_args(config_fs, data, action) {
     if (action === ACTIONS.ADD || action === ACTIONS.UPDATE) {
         if (action === ACTIONS.ADD) native_fs_utils.validate_bucket_creation({ name: data.name });
         if ((action === ACTIONS.UPDATE) && (data.new_name !== undefined)) native_fs_utils.validate_bucket_creation({ name: data.new_name });
-        if ((action === ACTIONS.ADD) && (data.bucket_owner === undefined)) throw_cli_error(ManageCLIError.MissingBucketOwnerFlag);
+        if ((action === ACTIONS.ADD) && (data.owner_account === undefined)) throw_cli_error(ManageCLIError.MissingBucketOwnerFlag);
```
Comment (shirady) [inline src/manage_nsfs/manage_nsfs_validations.js:319]: I think you can remove this line, we will never get here (we failed it before in the `get_bucket_owner_account`).

File: src/manage_nsfs/manage_nsfs_validations.js [hunk 17]
```diff
@@ -328,30 +328,33 @@ async function validate_bucket_args(config_fs, data, action) {
         if (!exists) {
             throw_cli_error(ManageCLIError.InvalidStoragePath, data.path);
         }
-        const account = await get_bucket_owner_account(config_fs, data.bucket_owner);
-        const account_fs_context = await native_fs_utils.get_fs_context(account.nsfs_account_config, data.fs_backend);
+
+        // bucket owner validations 
+        const owner_account = await get_bucket_owner_account(config_fs, undefined, data.owner_account);
```
Comment (shirady) [inline src/manage_nsfs/manage_nsfs_validations.js:333]: I think the variable name `owner_account` is a bit confusing because in the schema we have the same property name with string type (I understand that in the schema it should have been named `owner_account_id`... but we will probably not change it).

File: src/sdk/accountspace_fs.js [hunk 18]
```diff
@@ -26,6 +26,7 @@ const { account_cache } = require('./object_sdk');
 const dummy_region = 'us-west-2';
 const dummy_service_name = 's3';
 
+
```
Comment (shirady) [inline src/sdk/accountspace_fs.js:29]: redundent newline

File: src/cmd/manage_nsfs.js (hunk 14 above)
Comment (romayalon) [reply src/cmd/manage_nsfs.js:631]: we can say it also when inside the loop, we checked the first account but the account name was changed or new accounts were added... I think it's fine since it's list

File: src/manage_nsfs/manage_nsfs_validations.js (hunk 16 above)
Comment (romayalon) [reply src/manage_nsfs/manage_nsfs_validations.js:319]: I would keep it to be on the safe side, also since we know that these flows will hopefully change in the future, i'm not sure what will be the order then

File: src/cmd/manage_nsfs.js (hunk 14 above)
Comment (shirady) [reply src/cmd/manage_nsfs.js:631]: You are right, but a property of `bucket_owner` in a bucket might be not updated for years/months, unlike a bucket that was created during the loop which is not updated for seconds.

For example, the account and bucket were created in version 5.15 and in version 5.16 the account was renamed (bucket config was not updated), in version 5.18 when listing the buckets I think we should expect to see the bucket owner's updated name (and not the one which was originally saved).

Comment (romayalon) [reply src/cmd/manage_nsfs.js:631]: I think there is a missunderstnding, i'm not using the data.bucket_owner old property, inside set_bucket_owner(data) I read the account config by id and use it

File: src/cmd/manage_nsfs.js [hunk 19]
```diff
@@ -675,7 +675,7 @@ function get_access_keys(action, user_input) {
  */
 async function set_bucker_owner(bucket_data) {
     const account_data = await config_fs.get_identity_by_id(bucket_data.owner_account, TYPES.ACCOUNT, { silent_if_missing: true});
-    bucket_data.bucket_owner = account_data.name;
+    bucket_data.bucket_owner = account_data?.name;
```
Comment (shirady) [inline src/cmd/manage_nsfs.js:678]: I want to make sure I understand you - do you prefer to print undefined in the property `bucket_owner` than not printing it at all?

File: src/cmd/manage_nsfs.js (hunk 14 above)
Comment (shirady) [reply src/cmd/manage_nsfs.js:631]: Now I saw the map and the comment, thanks.

File: src/cmd/manage_nsfs.js (hunk 19 above)
Comment (romayalon) [reply src/cmd/manage_nsfs.js:678]: It removes undefined on the list, here is an example of a manual test - 
1. Create account `account1`
2. Create 11 buckets owned by `account1`
3. List buckets - see `bucket_owner: account1`
4. Delete manually on the FS `account1` identity
5. List buckets again - see `bucket_owner` is missing
```
"response": {
    "code": "BucketList",
    "reply": [
      {
        "_id": "66c59b3c40ae366b6414ffc5",
        "name": "bucket1",
        "owner_account": "66c59b21adca3d4be6b9f556",
        "versioning": "DISABLED",
        "creation_date": "2024-08-21T07:46:04.155Z",
        "path": "/private/tmp/dir2",
        "should_create_underlying_storage": false
      },
      {
        "_id": "66c59b579d6b1c9c554db214",
        "name": "bucket6",
        "owner_account": "66c59b21adca3d4be6b9f556",
        "versioning": "DISABLED",
        "creation_date": "2024-08-21T07:46:31.907Z",
        "path": "/private/tmp/dir2",
        "should_create_underlying_storage": false
      },
      {
    .....
      }
    ]
```

Comment (shirady) [reply src/cmd/manage_nsfs.js:678]: Thank you, I checked the flow in the code for this case:
list_config_files-> inside the bucket loop -> in the set_bucker_owner the bucker_owner is undefined -> in write_stdout_response inside it the `to_string` of list we will JSON.parse and JSON.stringify and this is what eliminates the `undefined` value as I understand it.

**Review by shirady — APPROVED** [review]
LGTM
